File: backend/lib/entities/order/order.rb
class Order
  attr_reader :customer, :items, :payment, :closed_at

  def initialize(customer, overrides = {})
    @customer = customer
    @items = []
    @order_item_class = overrides.fetch(:item_class) { OrderItem }
  end

  def add_product(product, quantity)
    @items << @order_item_class.new(order: self, product: product, quantity: quantity)
  end

  def set_payment(payment)
    @payment = payment

    close(payment.paid_at)
  end

  def total_amount
    @items.map(&:total).inject(:+) || 0
  end

private

  def close(closed_at = Time.now)
    @closed_at = closed_at
  end
end

File: backend/individual-assignment/challenge/src/main/kotlin/models/impl/Order.kt
package org.example.models.impl

import org.example.dtos.*
import org.example.enums.*
import org.example.extension.bookProcess
import org.example.extension.digitalProcess
import org.example.extension.membershipProcess
import org.example.extension.physicalProcess
import org.example.models.PaymentMethod
import java.util.*

class Order(
    val customer: Customer,
    val address: Address
) {
    private val items = mutableListOf<OrderItem>()
    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount
        get() = items.sumOf { it.total }

    fun addProduct(
        product: Product,
        quantity: Int
    ) {
        val productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception("The product have already been added. Change the amount if you want more.")

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null) {
            throw Exception("The order has already been paid!")
        }

        if (items.isEmpty()) {
            throw Exception("Empty order can not be paid!")
        }

        payment = Payment(this, method)

        close()
    }

    fun processProduct() {
        items.forEach{
            item ->
            when(item.product.type) {
                ProductType.BOOK -> bookProcess(
                    order = this
                )
                ProductType.PHYSICAL -> physicalProcess(
                    order = this
                )
                ProductType.DIGITAL -> digitalProcess(
                    order = this
                )
                ProductType.MEMBERSHIP -> membershipProcess(
                    order = this
                )
            }
        }
    }

    private fun close() {
        closedAt = Date()
    }
}